Add UpstreamOIDCProvider API and initial controller.

mattmoyer opened the pull request:
This adds a new `UpstreamOIDCProvider` in the `idp.supervisor.pinniped.dev/v1alpha1` API group.

This type describes an upstream OpenID Connect identity provider for the supervisor to target as an identity source:

For example:
```yaml
apiVersion: idp.supervisor.pinniped.dev/v1alpha1
kind: UpstreamOIDCProvider
metadata:
  name: test
  namespace: supervisor
spec:
  issuer: https://my-provider.com/oauth2/default
  client:
    secretName: test-client-secret
  authorizationConfig:
    redirectURI: https://localhost:12344/test-issuer/callback
    scopes:
    - openid
    - profile
    - email
  claims:
    username: email
    groups: groups
```

This PR does not necessarily reflect the final fields of this v1alpha1 API. We may still make small adjustments prior to the next numbered release.

There is also a new `upstream-watcher` controller which maintains the status conditions of each UpstreamOIDCProvider:

```cmd
$ kubectl get pinniped-idps -n supervisor
NAME   ISSUER                                       STATUS   AGE
test   https://dev-123.example.com/oauth2/default   Ready    2m10s

$ kubectl get pinniped-idp -n supervisor test -o yaml | yq read - status
conditions:
  - lastTransitionTime: "2020-11-12T00:36:28Z"
    message: loaded client credentials
    observedGeneration: 1
    reason: Success
    status: "True"
    type: ClientCredentialsValid
  - lastTransitionTime: "2020-11-12T00:34:37Z"
    message: discovered issuer configuration
    observedGeneration: 1
    reason: Success
    status: "True"
    type: OIDCDiscoverySucceeded
phase: Ready
```

This controller also performs OIDC discovery and updates a cache of validated OIDC upstream configurations, to be shared with the rest of the supervisor service.

**Release note**:

```release-note
Added new UpstreamOIDCProvider API for describing upstream OpenID Connect identity providers.
```

enj commented on internal/controller/supervisorconfig/upstreamwatcher/upstreamwatcher.go:
```diff
@@ -0,0 +1,310 @@
+	validatedUpstreams := make([]provider.UpstreamOIDCIdentityProvider, 0, len(actualUpstreams))
+	for _, upstream := range actualUpstreams {
+		if valid := c.validateUpstream(ctx, upstream); valid == nil {
+			overallError = errFailedToReconcile
```
`ErrSyntheticRequeue`

enj commented on internal/controller/supervisorconfig/upstreamwatcher/upstreamwatcher.go:
```diff
@@ -0,0 +1,310 @@
+func (c *controller) validateUpstream(ctx controllerlib.Context, upstream *v1alpha1.UpstreamOIDCProvider) *provider.UpstreamOIDCIdentityProvider {
+	result := provider.UpstreamOIDCIdentityProvider{
+		Name:   upstream.Name,
+		Scopes: upstream.Spec.AuthorizationConfig.Scopes,
```
add openid

enj commented on internal/controller/supervisorconfig/upstreamwatcher/upstreamwatcher.go:
```diff
@@ -0,0 +1,310 @@
+		}
+
+		// Update the cache with the newly discovered value.
+		_ = c.validatorCache.Set(upstream.Spec.Issuer, discoveredProvider)
```
just log in case

enj commented on internal/controller/supervisorconfig/upstreamwatcher/upstreamwatcher.go:
```diff
@@ -0,0 +1,310 @@
+	}
+
+	// Parse out and validate the discovered authorize endpoint.
+	authURL, err := url.Parse(discoveredProvider.Endpoint().AuthURL)
```
need checks for https

enj commented on hack/update.sh:
```diff
@@ -16,7 +16,7 @@ cp "$ROOT"/generated/1.19/crds/*.concierge.*.yaml "$ROOT/deploy/concierge"
 # Make sure we didn't miss any new CRDs.
 crdCount=$(find "$ROOT"/generated/1.19/crds/ -maxdepth 1 -type f -name '*.yaml' | wc -l | tr -d ' ')
-if [[ "$crdCount" != "3" ]]; then
+if [[ "$crdCount" != "4" ]]; then
```
I think this check can be removed.

cfryanr replied:
This check is to help us as developers make sure that we don't forget to update this script when we add a new CRD.

cfryanr replied:
Oh, I see that the script was updated to use glob matching to decide where the CRDs should get copied. Nevermind my above comment... it seems fine to remove this check.

mattmoyer replied:
Resolved in https://github.com/vmware-tanzu/pinniped/pull/204.

enj commented on test/integration/kube_api_discovery_test.go:
```diff
@@ -80,6 +80,39 @@ func TestGetAPIResourceList(t *testing.T) {
+						Namespaced:   true,
+						Kind:         "UpstreamOIDCProvider",
+						Verbs:        []string{"delete", "deletecollection", "get", "list", "patch", "create", "update", "watch"},
+						Categories:   []string{"all", "idp", "idps"},
```
assert our rules for categories and short names

enj commented on deploy/supervisor/rbac.yaml:
```diff
@@ -19,6 +19,12 @@ rules:
+    verbs: [get, list, watch, patch, update]
+  - apiGroups: [idp.supervisor.pinniped.dev]
+    resources: [upstreamoidcproviders/status]
+    verbs: [get, list, watch, patch, update]
```
remove list and watch

enj commented on internal/controller/supervisorconfig/upstreamwatcher/upstreamwatcher.go:
```diff
@@ -0,0 +1,310 @@
+	validatorCacheSize = 1024
+	validatorCacheTTL  = 1 * time.Hour
+
+	// Constants related to conditions.
+	typeClientCredsValid       = "ClientCredentialsValid"
+	typeOIDCDiscoverySucceeded = "OIDCDiscoverySucceeded"
+	reasonNotFound             = "SecretNotFound"
+	reasonWrongType            = "SecretWrongType"
+	reasonMissingKeys          = "SecretMissingKeys"
+	reasonSuccess              = "Success"
+	reasonUnreachable          = "Unreachable"
+	reasonInvalidResponse      = "InvalidResponse"
```
I am fine moving this.

enj commented on internal/controller/supervisorconfig/upstreamwatcher/upstreamwatcher_test.go:
```diff
@@ -0,0 +1,472 @@
+			}},
+		},
+		{
+			name: "existing valid upstream",
```
observed generation test

mattmoyer commented on Dockerfile:
```diff
@@ -25,6 +25,7 @@ RUN mkdir out \
 
 # Use a runtime image based on Debian slim
 FROM debian:10.6-slim
+RUN apt-get update && apt-get install -y ca-certificates && rm -rf /var/lib/apt/lists/*
```
We discussed this, but tl;dr for others -- it's so that the OIDC discovery can have access to a default CA bundle, because Debian `-slim` images don't have one.

mattmoyer commented on apis/supervisor/idp/doc.go.tmpl:
```diff
@@ -0,0 +1,8 @@
+// +k8s:deepcopy-gen=package
+// +groupName=idp.supervisor.pinniped.dev
+
+// Package idp is the internal version of the Pinniped supervisor identity provider (IDP) API.
```
This is a holdover from accidentally copy-pasting the structure from our aggregated APIs and `conversion-gen`. I'll file a separate PR to remove those across the board.

mattmoyer replied:
https://github.com/vmware-tanzu/pinniped/pull/202

mattmoyer commented on apis/supervisor/idp/v1alpha1/types_upstreamoidcprovider.go.tmpl:
```diff
@@ -0,0 +1,112 @@
+	// request flow with an OIDC identity provider.
+	// +kubebuilder:validation:Pattern=`^https://`
+	RedirectURI string `json:"redirectURI"`
+
```
Filed for later: #203

mattmoyer commented on apis/supervisor/idp/v1alpha1/types_upstreamoidcprovider.go.tmpl:
```diff
@@ -0,0 +1,112 @@
+// UpstreamOIDCProvider describes the configuration of an upstream OpenID Connect identity provider.
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:categories=all;idp;idps
```
Addressed for other APIs here: https://github.com/vmware-tanzu/pinniped/pull/205

mattmoyer commented:
@enj I think I addressed all the previous review comments, and I also switched the internal cache to use `k8s.io/apimachinery/pkg/util/cache` rather than `github.com/bluele/gcache`.

enj commented on internal/controller/supervisorconfig/upstreamwatcher/upstreamwatcher.go:
```diff
@@ -0,0 +1,320 @@
+// validateUpstream validates the provided v1alpha1.UpstreamOIDCProvider and returns the validated configuration as a
+// provider.UpstreamOIDCIdentityProvider. As a side effect, it also updates the status of the v1alpha1.UpstreamOIDCProvider.
+func (c *controller) validateUpstream(ctx controllerlib.Context, upstream *v1alpha1.UpstreamOIDCProvider) *provider.UpstreamOIDCIdentityProvider {
+	scopes := append([]string{"openid"}, upstream.Spec.AuthorizationConfig.AdditionalScopes...)
```
uh what happens if they ask for `openid` in additional scopes?

mattmoyer replied:
Oh, yeah maybe we should de-dupe as well. I think most upstream providers probably behave reasonably if you did send a duplicate scope, but we can eliminate that case pretty easily so we might as well.

enj commented on internal/controller/supervisorconfig/upstreamwatcher/upstreamwatcher.go:
```diff
@@ -0,0 +1,320 @@
+	}
+
+	// Validate the secret .data field.
+	_, hasID := secret.Data[clientIDDataKey]
+	_, hasSecret := secret.Data[clientSecretDataKey]
```
nit: store and use the values

enj reviewed (dismissed):
Minor nits
